src/cargo/core/compiler/fingerprint/mod.rs

```diff
@@ -772,10 +774,21 @@ impl LocalFingerprint {
     // TODO: This is allowed at this moment. Should figure out if it makes
     // sense if permitting to read env from the config system.
     #[allow(clippy::disallowed_methods)]
-    fn from_env<K: AsRef<str>>(key: K) -> LocalFingerprint {
+    fn from_env<K: AsRef<str>>(
+        key: K,
+        envs: &BTreeMap<String, Option<OsString>>,
+    ) -> LocalFingerprint {
         let key = key.as_ref();
         let var = key.to_owned();
-        let val = env::var(key).ok();
+
+        let val = envs
+            .get(key)
+            .and_then(|opt| {
+                opt.as_ref()
+                    .and_then(|os_str| os_str.clone().into_string().ok())
+            })
+            .or_else(|| env::var(key).ok());
```

Comment:
iiuc on Windows, `env::var` calls `GetEnvironmentVariableW` which is case insensitive.  Since we're recreating our own environment look up, should this matter?

Reply:
I don't have a more in-depth study of windows, I'm very sorry. 

Do you mean that you shouldn't use `env::var` or `env::var_os` to query environment variables? That doesn't seem to work : see: [rerun_if_env_changes](https://github.com/rust-lang/cargo/pull/14027/files#diff-bec1ea82298c9855dced8a66db57ca380f385e70254ca1cd3b8f99152465854dR55) and [rerun_if_env_or_file_changes](https://github.com/rust-lang/cargo/pull/14027/files#diff-bec1ea82298c9855dced8a66db57ca380f385e70254ca1cd3b8f99152465854dR109)